pkg/core/template/engine.go

```diff
@@ -259,6 +278,44 @@ func (e *Engine) resolveTranslations(templateType config.TemplateItemType) (tran
 	return
 }
 
+func (e *Engine) resolveComponents(types []config.TemplateItemType, key string) (bodies []string, err error) {
+	return e.resolveComponents0(types, key, make(map[config.TemplateItemType]struct{}))
+}
+
+func (e *Engine) resolveComponents0(types []config.TemplateItemType, key string, seen map[config.TemplateItemType]struct{}) (bodies []string, err error) {
```

Comment:
I think this can be a local function in `resolveComponents` to simplify it?
e.g.
```go
func (e *Engine) resolveComponents(types []config.TemplateItemType, key string) ([]string, error) {
    templates := map[config.TemplateItemType]string
    resolve := func(types []config.TemplateItemType) error {
        // ...
    }
    err := resolve(types)
    if err != nil {
        return nil, err
    }
    var bodies []string
    for _, body := range templates {
        bodies = append(bodies, body)
    }
    return bodies, nil
}
```